### PR TITLE
Even more UX Improvements

### PR DESCRIPTION
**Pull request by gakonst**
- [x] Add methods to deposit ETH/ERC20/ERC721 to User class, avoids this https://github.com/loomnetwork/plasma-cli/blob/master/src/index.ts#L106-L180
  - Adds ERC20/ERC721 ABIs
  - Adds method for ETH/ERC20/ERC721. ERC20 checks if allowance is enough, and increases it if not
- [x] Automatically triggers watching / stops watching on receiving a coin / starting an exit and on giving a coin away, avoid this https://github.com/loomnetwork/plasma-cli/blob/master/src/index.ts#L255-L270

### CODE REVIEW

File: src/plasma-cash/user.ts [hunk 1]
```diff
@@ -73,6 +81,53 @@ export class User extends Entity {
...
+  async depositETH(amount: BN): Promise<IPlasmaCoin> {
+    let currentBlock = await this.getCurrentBlockAsync()
+    const tx = await this.sendETH(this.plasmaCashContract._address, amount, 220000)
+    const coin = await this.logParser(tx, 0)
+    currentBlock = await this.pollForBlockChange(currentBlock, 20, 2000)
+    this.receiveAndWatchCoinAsync(coin.slot)
+    return coin
+  }
+
+  async depositERC721(uid: BN, address: string): Promise<IPlasmaCoin> {
```
Comment (enlight) [inline src/plasma-cash/user.ts]: `depositERC721Async`

File: src/plasma-cash/user.ts [hunk 2]
```diff
@@ -73,6 +81,53 @@ export class User extends Entity {
...
+      this.plasmaCashContract._address,
+      uid.toString()
+    ])
+    const coin = await this.logParser(tx, 1) // 2 events, transferred & deposited, we want the 2nd one
+    currentBlock = await this.pollForBlockChange(currentBlock, 20, 2000)
+    this.receiveAndWatchCoinAsync(coin.slot)
+    return coin
+  }
+
+  async depositERC20(amount: BN, address: string): Promise<IPlasmaCoin> {
```
Comment (enlight) [inline src/plasma-cash/user.ts]: `depositERC20Async`

File: src/plasma-cash/user.ts [hunk 3]
```diff
@@ -112,6 +168,67 @@ export class User extends Entity {
     return this.getCurrentBlockAsync()
   }
 
+  // Whenever a new block gets submitted refresh the user's state for their coins
+  async watchBlocks() {
```
Comment (enlight) [inline src/plasma-cash/user.ts]: `watchBlocks` itself doesn't need to be `async`

File: src/plasma-cash/entity.ts [hunk 4]
```diff
@@ -562,4 +596,20 @@ export class Entity {
       gas: this._defaultGas
     })
   }
+
+  /**
+   * 
+   * @param tx The transaction's receipt that wea want to decode
+   * @param i The Deposit event is the i'th emitted event. Set to 0 for depositing ETH, Set to 1 for ERC20/ERC721 since the first event is a `Transfer` event
+   */
+  async logParser(tx: any, i: number): Promise<IPlasmaCoin> {
```
Comment (enlight) [inline src/plasma-cash/entity.ts]: Maybe make this `getCoinFromTxAsync(txHash: string, i: number): Promise<IPlasmaCoin>`?

File: src/plasma-cash/entity.ts [hunk 5]
```diff
@@ -543,6 +561,22 @@ export class Entity {
     })
   }
 
+  async sendETH(to: string, value: BN, gas?: number): Promise<any> {
```
Comment (enlight) [inline src/plasma-cash/entity.ts]: `sendETHAsync`

File: src/plasma-cash/user.ts (hunk 3 above)
Comment (gakonst) [reply src/plasma-cash/user.ts]: I'm making some awaits inside, that's why

**Pull request merged**